### PR TITLE
[heos] Fix runtime error; Support for new  Denon "Home" speaker added

### PR DESCRIPTION
**Pull request by Wire82**
- This pull request fixes an runtime issue if no serial number is provided by the HEOS speaker. 
- It also adds the support to use the Denon "Home" speaker series as bridge

The runtime bug was reported within OpenHab [community](https://community.openhab.org/t/heos-denon-support/3449/380).

### CODE REVIEW

**Comment by martinvw** [comment]
@Wire82 does the serialnumber occur with a specific version of the software, or only with the new home speakers?

**Comment by Wire82** [comment]
> 
> 
> @Wire82 does the serialnumber occur with a specific version of the software, or only with the new home speakers?

@martinvw It seems that only the new Home speaker providing a serial number. The HEOS speaker (also the Denon Receiver with HEOS function) do not provide a serial number. I`m not able to say, if they disappeared with a software update in the past. 
So the runtime error only shows up for HEOS devices.

**Comment by martinvw** [comment]
> > @Wire82 does the serialnumber occur with a specific version of the software, or only with the new home speakers?
> 
> @martinvw It seems that only the new Home speaker providing a serial number. The HEOS speaker (also the Denon Receiver with HEOS function) do not provide a serial number. I`m not able to say, if they disappeared with a software update in the past.
> So the runtime error only shows up for HEOS devices.

The Representation property should be set for all discovered items, this makes a lot of sense but setting it to the same empty string for all devices is not going to solve the issue OH3 tries to warn us for.

Do you think we could use something else to identify them? Do we have a mac-address for them or something else unique? Maybe just fall back to the player-id. WDYT?

**Comment by Wire82** [comment]
> 
> 
> > > @Wire82 does the serialnumber occur with a specific version of the software, or only with the new home speakers?
> > 
> > 
> > @martinvw It seems that only the new Home speaker providing a serial number. The HEOS speaker (also the Denon Receiver with HEOS function) do not provide a serial number. I`m not able to say, if they disappeared with a software update in the past.
> > So the runtime error only shows up for HEOS devices.
> 
> The Representation property should be set for all discovered items, this makes a lot of sense but setting it to the same empty string for all devices is not going to solve the issue OH3 tries to warn us for.
> 
> Do you think we could use something else to identify them? Do we have a mac-address for them or something else unique? Maybe just fall back to the player-id. WDYT?

We could use the UDN identifier as already done for the bridge. But it's a very long number. If this isn't important we can use it. At the moment I was not able to get the MAC Address from the devices.

**Comment by martinvw** [comment]
> We could use the UDN identifier as already done for the bridge. But it's a very long number. If this isn't important we can use it.

Is that in tsmit's message the `dcbc810e-ea46-16cb-0080-0005cd51ce40` that does not look too long, seems like a perfect fallback to me.

**Comment by Wire82** [comment]
Yes that is the number. Okay, then I will use that number. I will check there it is used.

**Comment by martinvw** [comment]
I also had some pending changes, just created a PR :-)

I'll try to take another look after my lunch :-)

**Comment by Wire82** [comment]
I just noticed that this isn't that easy and I was wrong. The information about the players are received via the Telnnet connection. There, I don't get a UDN Identifier. This is only provided during the Upnp discovery. At the point where we need it, the only identifier unique available would the the player ID (PID).

**Comment by Wire82** [comment]
I tried it with the PID and this seems to work well.

**Comment by martinvw** [comment]
Do you happen to know whether that id is stable over time?

**Comment by Wire82** [comment]
Yes, they are. It happened one time, that the player IDs changed after a firmware upgrade. But Denon changed that with the next release. So therefore this will work. 
If we stay with the PID, shall I also reverse the change with the "" at the serial number?

File: bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/handler/HeosPlayerHandler.java [hunk 1]
```diff
@@ -179,6 +179,9 @@ public static void propertiesFromPlayer(Map<String, ? super String> prop, Player
         String serialNumber = player.serial;
         if (serialNumber != null) {
             prop.put(Thing.PROPERTY_SERIAL_NUMBER, serialNumber);
+        } else {
+            prop.put(Thing.PROPERTY_SERIAL_NUMBER, ""); // If no serial number is provided, write an empty string to
```
Comment (martinvw) [inline bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/handler/HeosPlayerHandler.java:183]: I would like to only use the `pid` as a fallback, so if the serialNumber is empty we use the `pid` on this line.

Comment (Wire82) [reply bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/handler/HeosPlayerHandler.java:183]: So using the PID as a fake serial number?

Comment (martinvw) [reply bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/handler/HeosPlayerHandler.java:183]: Yes

Comment (Wire82) [reply bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/handler/HeosPlayerHandler.java:183]: Okay. Good idea. I will change that and push the change.

**Review by martinvw — APPROVED** [review]
Great

**Comment by martinvw** [comment]
Sorry, I was/am a bit out of the loop, the build has succeeded but reports UNSTABLE what should we do in this case :-)

**Comment by Wire82** [comment]
Hmm I have no idea. I think it is not caused by the HEOS binding. I can try to trigger a new build job, to see if it will run through with the latest PRs included.

**Comment by martinvw** [comment]
@openhab/add-ons-maintainers what should we do with this output:

```
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:57 h (Wall Clock)
[INFO] Finished at: 2021-01-26T20:39:43Z
[INFO] ------------------------------------------------------------------------
Build timed out (after 240 minutes). Marking the build as aborted.
Build was aborted
channel stopped
```

Can I just merge?

**Comment by fwolter** [comment]
Seems to be a false positive. I see no reason not merging it.

**Comment by wborn** [comment]
I think @kaikreuzer changed the number of executors from 4 to 6 causing each build to take more than 4h now. It's already at 200% load when 4 executors build with 4 threads. So with 6 executors it is at 300% load.  Besides that it seems to have run out space too now. :thinking:

**Comment by kaikreuzer** [comment]
Oops, sorry for that - I'll reduce the number of executors again...

**Comment by Wire82** [comment]
@martinvw Thank you for merging this commit!

**Comment by martinvw** [comment]
> @martinvw Thank you for merging this commit!

More than welcome, thanks for your continued contribution!

**Pull request merged**